src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -131,6 +162,128 @@ public class FhirCosmosSearchService : SearchService
+                    return null;
+                }
+
+                string filteredType = expression.TargetResourceTypes.FirstOrDefault();
+                var includeExpressions = new List<IncludeExpression>();
+
+                if (expression.Reversed)
+                {
+                    // When reversed we'll use the Include expression code to return the ids
+                    // in the search index on the matched resources
+                    foreach (var targetInclude in expression.TargetResourceTypes)
+                    {
+                        includeExpressions.Add(Expression.Include(
+                            expression.ResourceTypes,
+                            expression.ReferenceSearchParameter,
+                            null,
+                            targetInclude,
+                            expression.TargetResourceTypes,
+                            false,
+                            false,
+                            false));
+                    }
+
+                    // When reversed the ids from the sub-query will match the base resource type
+                    filteredType = expression.ResourceTypes.FirstOrDefault();
+                }
+
+                MultiaryExpression filterExpression = Expression.And(
+                    Expression.SearchParameter(_resourceTypeSearchParameter, Expression.StringEquals(FieldName.TokenCode, null, filteredType, false)),
+                    criteria);
+
+                chainedOptions.Expression = filterExpression;
+
+                var chainedResults = await ExecuteSearchAsync<FhirCosmosResourceWrapper>(
+                    _queryBuilder.BuildSqlQuerySpec(chainedOptions, includeExpressions, idsOnly: true),
+                    chainedOptions,
+                    null,
+                    cancellationToken);
+
+                if (!string.IsNullOrEmpty(chainedResults.continuationToken))
+                {
+                    throw new InvalidSearchOperationException(string.Format(Resources.ChainedExpressionSubqueryLimit, _chainedSearchMaxSubqueryItemLimit));
+                }
+
+                Expression[] chainedExpressionReferences;
+
+                if (!expression.Reversed)
+                {
+                    // When normal chained expression we can filter using references in the parent object. e.g. Observation.subject
+                    // The following expression constrains "subject" references on "Observation" with the ids that have matched the sub-query
+                    chainedExpressionReferences = chainedResults.results.Select(x =>
+                            Expression.SearchParameter(
+                                expression.ReferenceSearchParameter,
+                                Expression.And(
+                                    Expression.Equals(FieldName.ReferenceResourceId, null, x.Id),
+                                    Expression.Equals(FieldName.ReferenceResourceType, null, filteredType))))
```

Comment:
>filteredType [](start = 93, length = 12)

Considering this can be null (it's a `FirstOrDefault()` and default is null for string) would it handle null correctly? Or we shouldn't add expression if it's a null?

Reply:
Changed to `First()`, it shouldn't be possible without a type